### PR TITLE
WIP: Cleanup core services during tests to fix memory leaks and flaky local tests - @mkhatib

### LINKED ISSUE #3138: Karma Tests are Memory Leaking causing tests to fail due to timeouts
After a lot of debugging of what's happening with the local tests, it seems that something is not being cleared correctly and hence karma runner keeps accumulating memory reaching up to 2GB! Causing our local machines to run out of memory and causing tests to run slower than they should. 

I believe this is the main cause of the local tests failing inconsistently with timeout errors. 

I am still trying to figure out what is leaking exactly and how to fix it. If anyone has ideas let me know!

### PR DESCRIPTION
**Pull request by mkhatib**
(please don't review details - still needs cleanup and some fixing)

There's still some more work to be done and still see a lot of objects not being garbage collected. 

Fixes #3138 #3000 

cc/ @dvoytenko @erwinmombay @jridgewell @sriramkrish85 

### Memory usage once all tests run - BEFORE change:
image

### Memory usage once all tests run - AFTER change:
image

### Doing 5000 dummy tests with test iframes - BEFORE change:
Reaches 1000+ tests - reachs 2GB - starts failing tests with timeouts and then crashes the tests after 1500+

### Doing 5000 dummy tests with test iframes - AFTER change:
Runs 5000 tests in ~7mins all successful and with 600MB memory upper limit at anytime and goes back to ~200MB

### CODE REVIEW

File: src/input.js [hunk 1]
```diff
@@ -84,14 +84,18 @@ export class Input {
...
+          win.document, 'mousemove', this.boundOnMouseMove_);
     }
   }
 
   /** @private */
   cleanup_() {
     this.win.document.removeEventListener('keydown', this.boundOnKeyDown_);
     this.win.document.removeEventListener('mousedown', this.boundOnMouseDown_);
+    if (this.unlistenToMouseMove_) {
+      this.unlistenToMouseMove_();
```
Comment (erwinmombay) [inline src/input.js]: you probly want to null out the prop afterwards

File: src/amp-core-service.js [hunk 2]
```diff
@@ -43,3 +43,16 @@ export function installCoreServices(window) {
   installUrlReplacementsService(window);
   installXhrService(window);
 }
+
+export function uninstallCoreServices(window) {
```
Comment (erwinmombay) [inline src/amp-core-service.js]: should probly be marked `@visibleForTesting` only

**Comment by cramforce** [comment]
Lets talk about this in person. This is a great change! But we will need to make some adjustments. I also believe @jridgewell might have looked into similar things.

- None of the `this.foo = null` changes should be necessary except in rare cases and we should not do this unless absolutely needed.
- Unregisterung a service should not typically be needed on iframes that get destroyed anyway. Just destroy the iframe.
- Removing the iframes created for each tests is obviously the right thing. Unfortunately I've never been able to get this to pass in Safari. Possibly your other changes fix this.
- Unlistening global events should not be necessary, if we remove the iframes. This, might, however help with Safari.

**Comment by jridgewell** [comment]
Re: https://github.com/ampproject/amphtml/pull/2596, I updated our test dependencies to a new version of Mocha that solved the [memory leak](https://github.com/mochajs/mocha/commit/33c825ba054fc4dfdf6e270ec8369dd4a28530e1).

**Comment by mkhatib** [comment]
Thanks @jridgewell and @cramforce 
> Re: #2596, I updated our test dependencies to a new version of Mocha that solved the memory leak.
Thanks for bringing this to my attention I totally missed it. As discussed this seems to help with reclaiming some of the memory once the tests finish but doesn't help with the fast growth of memory and tests will still crash and timeout. Reaching up to 2.2GB and idling out at 1.2GB.

Removing `iframes` (as in _init_test afterEach cleanup) with nothing else in this change helps slow the growth but would not be enough. This seems to allow the browser to GC more objects, lowering down the growth by almost 50% by the end of the tests to 1.2GB and idling out at around 600MB after the tests. This would probably fix our current problem with tests timing out BUT this will resurface the more tests we add, so really delay the issue rather than fixes it.

> Removing the iframes created for each tests is obviously the right thing. Unfortunately I've never been able to get this to pass in Safari. Possibly your other changes fix this.
Is this `gulp test --integration --safari` or `gulp test --safari`? The first seems to pass (now it's actually back to failing, will debug some more), but the second seems to be failing (even without this change) with different errors, but the most common is:

```sh
	TypeError: win.document.registerElement is not a function. (In 'win.document.registerElement(name, {
	    prototype: createAmpElementProto(win, name, implementationClass)
	  })', 'win.document.registerElement' is undefined) (/var/folders/7w/bvjck7ls7f58m040x_qvph4w002y5_/T/24a18b56928e971433d6dda2ca27792e.browserify:41565 <- /Users/mkhatib/amphtml/src/custom-element.js:1220:2)
Safari 9.1.0 (Mac OS X 10.11.4): Executed 2 of 2026 (1 FAILED) (0 secs / NaN secs)
```

Thanks for all the comments, I would like to clean it up some more and these would help.

**Pull request closed without merging**